### PR TITLE
Windows support: switch from TCP transport to Named Pipes for the Workload API 

### LINKED ISSUE #2932: Windows support: switch from TCP transport to Named Pipes for the Workload API
We have found some limitations using the Workload API with TCP transport that makes it impractical to use in containerized environments.
Additionally, any endpoint that requires both PID resolution and be protected by ACL (like the agent Admin API) would be very problematic to implement with a TCP transport.

### PR DESCRIPTION
**Pull request by amartinezfayo**
Switch from TCP transport to Named Pipes for the Workload API in Windows.

Fixes #2932.

### CODE REVIEW

File: cmd/spire-agent/cli/run/run_windows.go [hunk 1]
```diff
@@ -30,7 +31,7 @@ func (c *agentConfig) getAdminAddr() (*net.UnixAddr, error) {
 // validateOS performs windows specific validations of the agent config
 func (c *agentConfig) validateOS() error {
 	if c.SocketPath != "" {
-		return errors.New("invalid configuration: socket_path is not supported in this platform; please use tcp_socket_port instead")
+		return errors.New("invalid configuration: socket_path is not supported in this platform; please use named_pipe_socket_path instead")
```
Comment (azdagron) [inline cmd/spire-agent/cli/run/run_windows.go:34]: named_pipe_name?

File: pkg/agent/api/endpoints_windows.go [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+package api
+
+import (
+	"net"
+
+	"github.com/spiffe/spire/pkg/common/peertracker"
+)
+
+func (e *Endpoints) createListener() (net.Listener, error) {
+	return nil, peertracker.ErrUnsupportedPlatform
```
Comment (azdagron) [inline pkg/agent/api/endpoints_windows.go:13]: This is coming in with a follow-up PR, right?

File: pkg/common/peertracker/npipe_windows.go [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+	}
+
+	return CallerInfo{
+		Addr: conn.RemoteAddr(),
+		PID:  pid,
+	}, nil
+}
+
+func getNamedPipeClientProcessID(pipe windows.Handle, clientProcessID *int32) (err error) {
+	r1, _, e1 := syscall.Syscall(procGetNamedPipeClientProcessID.Addr(), 2, uintptr(pipe), uintptr(unsafe.Pointer(clientProcessID)), 0)
```
Comment (azdagron) [inline pkg/common/peertracker/npipe_windows.go:48]: Just a heads up that this will need to change to the new SyscallN function when we move to go1.18.

**Review by azdagron — COMMENTED** [review]
A thing of beauty! Just had a few comments.

File: pkg/agent/api/endpoints_windows.go (hunk 2 above)
Comment (amartinezfayo) [reply pkg/agent/api/endpoints_windows.go:13]: Yes, we have https://github.com/spiffe/spire/issues/2833 filed for this.

File: pkg/common/peertracker/npipe_windows.go [hunk 4]
```diff
@@ -55,7 +57,7 @@ func getNamedPipeClientProcessID(pipe windows.Handle, clientProcessID *int32) (e
 // compareObjectHandles compares two object handles to determine if they
 // refer to the same underlying kernel object
 func compareObjectHandles(firstHandle, secondHandle windows.Handle) error {
-	r1, _, e1 := syscall.SyscallN(procCompareObjectHandles.Addr(), uintptr(firstHandle), uintptr(secondHandle))
+	r1, _, e1 := syscall.Syscall(procCompareObjectHandles.Addr(), 2, uintptr(firstHandle), uintptr(secondHandle), 0)
```
Comment (azdagron) [inline pkg/common/peertracker/npipe_windows.go:60]: Don't we still want to use SyscallN here?

File: pkg/common/peertracker/npipe_windows.go [hunk 5]
```diff
@@ -44,8 +44,10 @@ func getCallerInfoFromNamedPipeConn(conn net.Conn) (CallerInfo, error) {
 	}, nil
 }
 
+// getNamedPipeClientProcessID retrieves the client process identifier
+// for the specified handle representing a named pipe.
 func getNamedPipeClientProcessID(pipe windows.Handle, clientProcessID *int32) (err error) {
-	r1, _, e1 := syscall.Syscall(procGetNamedPipeClientProcessID.Addr(), 2, uintptr(pipe), uintptr(unsafe.Pointer(clientProcessID)), 0)
+	r1, _, e1 := syscall.SyscallN(procGetNamedPipeClientProcessID.Addr(), 2, uintptr(pipe), uintptr(unsafe.Pointer(clientProcessID)), 0)
```
Comment (azdagron) [inline pkg/common/peertracker/npipe_windows.go:50]:
```suggestion
	r1, _, e1 := syscall.SyscallN(procGetNamedPipeClientProcessID.Addr(), uintptr(pipe), uintptr(unsafe.Pointer(clientProcessID)))
```

File: pkg/common/peertracker/npipe_windows.go (hunk 4 above)
Comment (amartinezfayo) [reply pkg/common/peertracker/npipe_windows.go:60]: Oops, missed that, thanks!

File: pkg/common/peertracker/npipe_windows.go (hunk 5 above)
Comment (azdagron) [reply pkg/common/peertracker/npipe_windows.go]: This use of SyscallN is incorrect. SyscallN just takes a slice of args and determines the correct number of args using that.

**Review by azdagron — APPROVED** [review]
\o/!

**Pull request merged**